src/include/daos_cont.h

```diff
@@ -55,10 +58,10 @@ typedef struct {
 	uint32_t		ci_redun_fac;
 	/** Number of snapshots */
 	uint32_t		ci_nsnapshots;
-	/** Redundancy level */
-	uint32_t		ci_redun_lvl;
-	/** Container information pad (not used) */
-	uint32_t		ci_pad[3];
+	/** Latest open time (hybrid logical clock) */
+	uint64_t		ci_md_otime;
+	/** Latest close/modify time (hybrid logical clock) */
+	uint64_t		ci_md_mtime;
```

Comment:
Good question, I'm not sure. @mjmac recently commited https://github.com/daos-stack/daos/commit/c5e3363321a3f83883d39a2ff49308ef5fa4f6d3 to abstract HLC and present the ci_lsnapshot value in human-readable form  (for interactive daos utility users). I had been thinking about applying that same approach for the (HLC) metadata times in a subsequent patch for this work.

For API users in particular perhaps it would be better to report the metadata times as Unix second timestamps (while keeping ci_lsnapshot as an epoch / HLC value)? Let me think about this some more while I continue the work. For this patch I guess I'll keep metadata times as HLC values. Before 2.4 is released perhaps I might change them to Unix seconds. I know that one of the API use cases (likely to be used first by tools like datamover) will be in the pool API to list/filter containers by (e.g., metadata time) search criteria: e.g., "show me the list of containers in the pool with close_modify time older than 30 days"). It seems such callers might have some difficulty working with HLC values?

Reply:
just FYI, i recently added an hlc2timespec conversion function:
https://github.com/daos-stack/daos/blob/master/src/include/cart/api.h#L545
if you think it is useful here.